---
title: "scAPAmod: profiling alternative polyadenylation modalities in single cells from single-cell RNA-seq data"
author: "Xiaohui Wu, Hongjuan Fu"
date: "2021-05-20"
output:
  pdf_document:
    latex_engine: xelatex
    toc:  true
    number_sections: true
    # df_print: kable
urlcolor: blue
fontsize: 11pt
vignette: >
  %\VignetteIndexEntry{scAPAmod}
  %\VignetteEngine{knitr::rmarkdown}
  %\VignetteEncoding{UTF-8}
---

```{r setup, include=FALSE}
knitr::opts_chunk$set(
  fig.width = 6,
  fig.height = 5.5,
  collapse = TRUE,
  warning = FALSE,
  comment = ""
)
```

# Preparations
## PAC data of mouse sperm cells
```{r message=FALSE}
library(movAPA, warn.conflicts = FALSE, quietly=TRUE)
data(scPACds)
head(scPACds@counts[1:2,1:5]) 
head(scPACds@anno, n=2)
head(scPACds@colData, n=2) 
levels(scPACds@colData$celltype)
```
## Preprocess
```{r message=FALSE}
library(scAPAmod, warn.conflicts = FALSE, quietly=TRUE)
# 3' UTR
index <- which(scPACds@anno$ftr == "3UTR")
UTR_gene <- scPACds@anno$gene[index]
UTR_chr <- scPACds@anno$chr[index]
UTR_strand <- scPACds@anno$strand[index]
UTR_coord <- scPACds@anno$coord[index]
UTR_ftr_start <- scPACds@anno$ftr_start[index]
UTR_ftr_end <- scPACds@anno$ftr_end[index]
UTR_three_UTR_length <- scPACds@anno$three_UTR_length[index]
UTR_anno <- data.frame(chr = as.character(UTR_chr),
                       strand = as.character(UTR_strand),
                       coord =as.integer(UTR_coord),
                       gene = as.character(UTR_gene),
                       ftr_start = as.integer(UTR_ftr_start),
                       ftr_end = as.integer(UTR_ftr_end),
                       three_UTR_length = as.integer(UTR_three_UTR_length))
UTR_counts <-  scPACds@counts[,index]
ct1 <- which(scPACds@colData$celltype[index] == "SC")
result1 <- extrPairPA(UTR_counts[,ct1],
                    as.character(UTR_anno$gene),UTR_anno)
# non 3' UTR
ct <- which(scPACds@colData$celltype == "SC")
results <- exnon3UTRPA(scPACds@counts[,ct],
                       scPACds@anno$gene, scPACds@anno, scPACds@anno$ftr,
                       gn = 1, cn = 1)
```
# Analyses of APA dynamics
## Identifying modalities in 3' UTR
```{r message=FALSE}
mod <- getMod(result1$PUI)
mod$modalities
```
if you want to see the modalities directly in 3' UTR, you can use *UTRmod*
```{r message=FALSE}
scUTRds <- get3UTRAPAds(scPACds)
mod <- UTRmod(scUTRds,ct1)
mod$modalities
```
## Identifying modalities in non 3' UTR
```{r message=FALSE}
nonmod <- getMod(results$PUI)
nonmod$modalities
# use chi-square test to test Bimodal
ind <- which(nonmod$modalities == "Bimodal")
bigene <- results$gene[ind]
label <- lapply(c(1:length(nonmod$results)), function(y){
  la <- nonmod$results[[y]][[2]][["cluster_labels"]]
  if (length(which(is.na(results$PUI[y,])==TRUE))>0) {
    dat.tmp <- results$PUI[y,][-which(is.na(results$PUI[y,]))]
  }else{
    dat.tmp <- results$PUI[y,]
  }
  names(la) <- names(dat.tmp)
  return(la)})
bilabel <- label[ind]

# if(length(which(is.na(bigene)))>0){
#   id <- which(is.na(bigene))
#   bigene <- bigene[-id]
#   bilabel <- bilabel[-id]}
pval1 <- chisqtest(results$filter.data, results$gene, bigene, bilabel)
pval1 <- p.adjust(pval1, method = "BH")
# use KS test to test Bimodal
pval2 <- KStest(results$PUI, results$gene, results$ftr, bigene, bilabel)
pval2 <- p.adjust(pval2, method = "BH")
```
if you want to see the modalities directly in non 3' UTR, you can use *nonUTRmod*
```{r message=FALSE}
nonmod <- nonUTRmod(scPACds,ct,gn = 1, cn = 1)
nonmod$modalities
```
## Research APA preferences
There are two types of APA preferences, one is the major PA, and the other is the minor PA. Here, the ratio value of the APA data is calculated, and the largest is extracted as the major PA, and the smallest is the minor PA. Analyze the modalities of APA usage with different preferences and the distribution of APA usage modalities in different regions.
```{r message=FALSE}
# the major PA
mmod <- MAMIMod(scPACds,"SC","MajorPA")
mmod$modalities
# the minor PA
mimod <- MAMIMod(scPACds,"SC","MinorPA")
mimod$modalities
```
If you want to see the ratio value specifically, you can use *exMajorPA*, or if you want to directly model the recognition modalities, you can use *getMMod* which is a bit different from PUI data modeling.
```{r message=FALSE}
# the major PA and minor PA
mresult <- exMajorPA(scPACds,"SC")
# modalities of majorPA
mmod <- getMMod(mresult$PAmax,"PAmax")
mmod$modalities
# modalities of minorPA
mimod <- getMMod(mresult$PAmin,"PAmin")
mimod$modalities
```

# Statistics of modalities
## Statistics on modalities distribution of different cell types
```{r message=FALSE}
# cell type of RS
ct2 <- which(scPACds@colData$celltype[index] == "RS")
result2 <- extrPairPA(UTR_counts[,ct2],
                    as.character(UTR_anno$gene),UTR_anno)
mod2 <- getMod(result2$PUI)
mod2$modalities
# cell type of ES
ct3 <- which(scPACds@colData$celltype[index] == "ES")
result3 <- extrPairPA(UTR_counts[,ct3],
                    as.character(UTR_anno$gene),UTR_anno)
mod3 <- getMod(result3$PUI)
mod3$modalities
```
```{r message=FALSE}
# set the cell type
celltype <- c(rep("SC", 3), rep("RS", 3), rep("ES", 3))
data <- data.frame(celltype)
data$modality <- c("Bimodal","Multimodal","Unimodal")
data$number <- c(table(mod$modalities),table(mod2$modalities),table(mod3$modalities))
ggplot(data, aes(x=celltype, y=number)) +
  ggplot2::geom_bar(stat = "identity", position = "dodge", aes(fill=modality))

```


## Distribution of usage modalties at different stages of differentiation
```{r message=FALSE, fig.showtext=TRUE}
library(ggplot2)
library(ggstatsplot)
#extrafont::loadfonts()
data("tUTRModalChange")
ggstatsplot::ggbarstats(data = staChange, x = condition, y = celltype,
                        title = "Exchange of modalities from different cell type",
                        ylab = "% of total common gene of PA",
                        # ggstatsplot.layer = FALSE,
                        sampling.plan = "jointMulti",
                        ggtheme = hrbrthemes::theme_ipsum_pub(),
                        legend.title = "condition", messages = F, palette = "Set2")

```

## Changes in the modalities of different cell differentiation stages
```{r message=FALSE, fig.showtext=TRUE}
library(extrafont)
data("tUTRModalChangeDetail")
ggstatsplot::ggbarstats(data = detChange, x = condition, y = celltype,
                        title = "Exchange of modalities from different cell type",
                        ylab = "% of total common gene of PA",
                        # ggstatsplot.layer = FALSE,
                        sampling.plan = "jointMulti",
                        ggtheme = hrbrthemes::theme_ipsum_pub(),
                        legend.title = "condition", messages = F, palette = "Set2")

```

## Visualize the distribution of PA expression according to the components
```{r message=FALSE}
# heatmap
library(grid)
library(org.Mm.eg.db)
PUI <- result1$PUI
tUTR.pair.cd.tmp <- result1$filter.data
tUTR.gene <- result1$gene
tUTR.gene.pui <- rownames(PUI)
which(mod$modalities == "Bimodal")
id2 <- tUTR.gene.pui[3]
genename <- select(org.Mm.eg.db, keys = id2, 
                   columns = c("SYMBOL","ENTREZID","GENENAME"),
                   keytype = "ENSEMBL")
tUTR.gene <- select(org.Mm.eg.db, keys = tUTR.gene, 
                    columns = c("SYMBOL","ENTREZID","GENENAME"),
                    keytype = "ENSEMBL")
plotGenePACount3(org.Mm.eg.db, genename$ENTREZID, tUTR.pair.cd.tmp,
                 tUTR.gene$ENTREZID,label[[3]])

```

## GO analysis
```{r message=FALSE,fig.height=4, fig.width=8, size=8}
library(clusterProfiler)
scego <- enrichGO(OrgDb="org.Mm.eg.db", gene = rownames(result1$PUI),
                  keyType = "ENSEMBL", ont = "ALL", pAdjustMethod = "BH",
                  pvalueCutoff = 0.05,qvalueCutoff = 0.05, readable= TRUE)

barplot(scego,showCategory = 10)
```

# Session Information
The session information records the versions of all the packages used in the generation of the present document.
```{r}
sessionInfo()
```
